File: WinForms/CS/ChartsMainDemo/Modules/ChartViews/ChartDemoModuleLabelViews.cs
using System;
using DevExpress.Utils;
using DevExpress.Utils.Design;
using DevExpress.XtraBars.Docking;
using DevExpress.XtraBars.Docking2010.Views;
using DevExpress.XtraBars.Docking2010.Views.Tabbed;
using DevExpress.XtraEditors;
using DevExpress.XtraLayout;

namespace DevExpress.XtraCharts.Demos {
    public class ChartDemoModuleLabelViews : ChartDemoModuleWithOptions {
        protected const string PiePercentPattern = "{A}: {VP:P2}";
        protected const string FunnelPercentPattern = "{A}: {VP:P0}";
        protected const string PercentPattern = "{VP:P0}";

        ChartControl activeChart = new ChartControl();

        protected virtual SeriesBase ActiveSeries {
            get {
                if(ChartControl.Series.Count > 0)
                    return ChartControl.Series[0];
                return ChartControl.SeriesTemplate;
            }
        }
        protected virtual SpinEdit LabelAngleSpinEdit {
            get { return null; }
        }
        protected virtual CheckEdit ValueAsPercentCheckEdit {
            get { return null; }
        }
        protected virtual CheckEdit LabelVisibleCheckEdit {
            get { return null; }
        }
        protected virtual LayoutControl OptionsLayoutControl {
            get { return null; }
        }
        protected virtual DefaultBoolean CrosshairEnabled {
            get { return DefaultBoolean.Default; }
        }
        protected virtual DefaultBoolean ToolTipEnabled {
            get { return DefaultBoolean.Default; }
        }
        protected virtual bool AnimationEnabled {
            get { return true; }
        }
        protected virtual TabbedView TabbedView {
            get { return null; }
        }

        internal override ChartControl ChartControl {
            get { return activeChart; }
        }

        public ChartDemoModuleLabelViews() {
            Load += ChartDemoModuleLabelViews_Load;
        }

        void ChartDemoModuleLabelViews_Load(object sender, EventArgs e) {
            if(!DesignTimeTools.IsDesignMode)
                InitControls();
        }
        bool IsFullStackedBarView(SeriesViewBase view) {
            return view is FullStackedBarSeriesView || view is FullStackedBar3DSeriesView;
        }
        bool IsSideBySideFullStackedBarView(SeriesViewBase view) {
            return view is SideBySideFullStackedBarSeriesView || view is SideBySideFullStackedBar3DSeriesView;
        }
        bool IsFullStackedView(SeriesViewBase view) {
            return view is FullStackedAreaSeriesView ||
                   view is FullStackedSplineAreaSeriesView ||
                   view is FullStackedLineSeriesView ||
                   view is FullStackedArea3DSeriesView ||
                   view is FullStackedLine3DSeriesView;
        }
        bool IsFunnelView(SeriesViewBase view) {
            return view is FunnelSeriesView || view is Funnel3DSeriesView;
        }
        bool IsPieView(SeriesViewBase view) {
            return view is PieSeriesView || view is Pie3DSeriesView;
        }
        void ValueAsPercentCheckedEditCheckedChanged(object sender, EventArgs e) {
            ApplySettingsToSeries(SetValueAsPercent);
        }
        void LabelAngleSpinEditEditValueChanged(object sender, EventArgs e) {
            ApplySettingsToSeries(SetLabelAngle);
        }
        void OnLabelVisibilityChanged(LayoutControlItem[] items) {
            bool labelVisible = LabelVisibleCheckEdit.Checked;
            SetEnabledLayoutControlItems(labelVisible, items);
            ChartControl.CrosshairEnabled = labelVisible ? DefaultBoolean.False : CrosshairEnabled;
            ChartControl.ToolTipEnabled = labelVisible ? DefaultBoolean.False : ToolTipEnabled;
        }
        void UpdateControls() {
            if(OptionsLayoutControl == null)
                return;
            OptionsLayoutControl.BeginUpdate();
            UpdateControlsCore();
            OptionsLayoutControl.EndUpdate();
        }
        void TabbedView_PopupMenuShowing(object sender, XtraBars.Docking2010.Views.PopupMenuShowingEventArgs e) {
            e.Cancel = e.GetDocument() != null;
        }

        protected virtual void InitControls() {
            if(TabbedView != null)
                TabbedView.PopupMenuShowing += TabbedView_PopupMenuShowing;
            SetDataSources();
            if(LabelAngleSpinEdit != null)
                LabelAngleSpinEdit.EditValueChanged += LabelAngleSpinEditEditValueChanged;
            if(ValueAsPercentCheckEdit != null)
                ValueAsPercentCheckEdit.CheckedChanged += ValueAsPercentCheckedEditCheckedChanged;
        }
        protected virtual void SetDataSources() { }
        protected virtual void UpdateControlsCore() { }
        protected virtual void OnDocumentActivatedCore(BaseDocument document) {
            UpdateControls();
            if(AnimationEnabled)
                activeChart.Animate();
        }
        protected void ApplySettingsToSeries(Action<SeriesBase> action) {
            if(string.IsNullOrEmpty(ChartControl.SeriesDataMember) && ChartControl.SeriesTemplate.DataAdapter == null)
                foreach(Series series in ChartControl.Series)
                    action(series);
            else
                action(ChartControl.SeriesTemplate);
        }
        protected void ApplySettingsToSeriesView(Action<SeriesViewBase> viewAction) {
            ApplySettingsToSeries((series) => { viewAction(series.View); });
        }
        protected void SetEnabledLayoutControlItems(bool enabled, params LayoutControlItem[] items) {
            foreach(LayoutControlItem item in items)
                item.Enabled = enabled;
        }
        protected void OnDocumentActivated(BaseDocument document) {
            if(!document.IsDockPanel)
                return;
            activeChart = ((DockPanel)document.Control).ControlContainer.Controls[0] as ChartControl;
            OnDocumentActivatedCore(document);
        }
        protected void SetLabelAngle(SeriesBase series) {
            SeriesLabelBase label = series.Label;
            int labelAngle = (int)LabelAngleSpinEdit.Value;
            PointSeriesLabel pointLabel = label as PointSeriesLabel;
            if(pointLabel != null) {
                pointLabel.Angle = labelAngle;
                return;
            }
            RangeAreaSeriesLabel rangeAreaLabel = label as RangeAreaSeriesLabel;
            if(rangeAreaLabel != null) {
                rangeAreaLabel.MinValueAngle = labelAngle;
                rangeAreaLabel.MaxValueAngle = labelAngle;
            }
        }
        protected void SetLabelVisibility(params LayoutControlItem[] items) {
            OnLabelVisibilityChanged(items);
            DefaultBoolean labelsVisibility = Utils.CovertBoolToDefaultBoolean(LabelVisibleCheckEdit.Checked);
            ApplySettingsToSeries((series) => series.LabelsVisibility = labelsVisibility);
        }
        protected void SetLabelVisibility(SeriesBase series, params LayoutControlItem[] items) {
            OnLabelVisibilityChanged(items);
            series.LabelsVisibility = Utils.CovertBoolToDefaultBoolean(LabelVisibleCheckEdit.Checked);
        }
        protected void SetValueAsPercent(SeriesBase series) {
            SeriesViewBase view = series.View;
            bool valueAsPercent = ValueAsPercentCheckEdit.Checked;
            if(view is FullStackedStepAreaSeriesView)
                series.Label.TextPattern = valueAsPercent ? PercentPattern : "{V:F0}";
            else if(IsPieView(view))
                series.Label.TextPattern = valueAsPercent ? PiePercentPattern : "{A}: {V:F1}";
            else if(IsFunnelView(view))
                series.Label.TextPattern = valueAsPercent ? FunnelPercentPattern : "{A}: {V:F0}";
            else if(IsSideBySideFullStackedBarView(view))
                series.Label.TextPattern = valueAsPercent ? PercentPattern : "{V:0,,.0}M";
            else if(IsFullStackedBarView(view))
                series.Label.TextPattern = valueAsPercent ? PercentPattern : "{V:F0}K";
            else if(IsFullStackedView(view))
                series.Label.TextPattern = valueAsPercent ? PercentPattern : "${V}M";
        }
    }
}

File: WinForms/VB/ChartsMainDemo/Modules/ChartViews/ChartDemoModuleLabelViews.vb
Imports System
Imports DevExpress.Utils
Imports DevExpress.Utils.Design
Imports DevExpress.XtraBars.Docking
Imports DevExpress.XtraBars.Docking2010.Views
Imports DevExpress.XtraBars.Docking2010.Views.Tabbed
Imports DevExpress.XtraEditors
Imports DevExpress.XtraLayout

Namespace DevExpress.XtraCharts.Demos

    Public Class ChartDemoModuleLabelViews
        Inherits ChartDemoModuleWithOptions

        Protected Const PiePercentPattern As String = "{A}: {VP:P2}"

        Protected Const FunnelPercentPattern As String = "{A}: {VP:P0}"

        Protected Const PercentPattern As String = "{VP:P0}"

        Private activeChart As ChartControl = New ChartControl()

        Protected Overridable ReadOnly Property ActiveSeries As SeriesBase
            Get
                If ChartControl.Series.Count > 0 Then Return ChartControl.Series(0)
                Return ChartControl.SeriesTemplate
            End Get
        End Property

        Protected Overridable ReadOnly Property LabelAngleSpinEdit As SpinEdit
            Get
                Return Nothing
            End Get
        End Property

        Protected Overridable ReadOnly Property ValueAsPercentCheckEdit As CheckEdit
            Get
                Return Nothing
            End Get
        End Property

        Protected Overridable ReadOnly Property LabelVisibleCheckEdit As CheckEdit
            Get
                Return Nothing
            End Get
        End Property

        Protected Overridable ReadOnly Property OptionsLayoutControl As LayoutControl
            Get
                Return Nothing
            End Get
        End Property

        Protected Overridable ReadOnly Property CrosshairEnabled As DefaultBoolean
            Get
                Return DefaultBoolean.Default
            End Get
        End Property

        Protected Overridable ReadOnly Property ToolTipEnabled As DefaultBoolean
            Get
                Return DefaultBoolean.Default
            End Get
        End Property

        Protected Overridable ReadOnly Property AnimationEnabled As Boolean
            Get
                Return True
            End Get
        End Property

        Protected Overridable ReadOnly Property TabbedView As TabbedView
            Get
                Return Nothing
            End Get
        End Property

        Friend Overrides ReadOnly Property ChartControl As ChartControl
            Get
                Return activeChart
            End Get
        End Property

        Public Sub New()
            AddHandler Load, AddressOf ChartDemoModuleLabelViews_Load
        End Sub

        Private Sub ChartDemoModuleLabelViews_Load(ByVal sender As Object, ByVal e As EventArgs)
            If Not DesignTimeTools.IsDesignMode Then InitControls()
        End Sub

        Private Function IsFullStackedBarView(ByVal view As SeriesViewBase) As Boolean
            Return TypeOf view Is FullStackedBarSeriesView OrElse TypeOf view Is FullStackedBar3DSeriesView
        End Function

        Private Function IsSideBySideFullStackedBarView(ByVal view As SeriesViewBase) As Boolean
            Return TypeOf view Is SideBySideFullStackedBarSeriesView OrElse TypeOf view Is SideBySideFullStackedBar3DSeriesView
        End Function

        Private Function IsFullStackedView(ByVal view As SeriesViewBase) As Boolean
            Return TypeOf view Is FullStackedAreaSeriesView OrElse TypeOf view Is FullStackedSplineAreaSeriesView OrElse TypeOf view Is FullStackedLineSeriesView OrElse TypeOf view Is FullStackedArea3DSeriesView OrElse TypeOf view Is FullStackedLine3DSeriesView
        End Function

        Private Function IsFunnelView(ByVal view As SeriesViewBase) As Boolean
            Return TypeOf view Is FunnelSeriesView OrElse TypeOf view Is Funnel3DSeriesView
        End Function

        Private Function IsPieView(ByVal view As SeriesViewBase) As Boolean
            Return TypeOf view Is PieSeriesView OrElse TypeOf view Is Pie3DSeriesView
        End Function

        Private Sub ValueAsPercentCheckedEditCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            ApplySettingsToSeries(New Action(Of SeriesBase)(AddressOf SetValueAsPercent))
        End Sub

        Private Sub LabelAngleSpinEditEditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            ApplySettingsToSeries(New Action(Of SeriesBase)(AddressOf SetLabelAngle))
        End Sub

        Private Sub OnLabelVisibilityChanged(ByVal items As LayoutControlItem())
            Dim labelVisible As Boolean = LabelVisibleCheckEdit.Checked
            SetEnabledLayoutControlItems(labelVisible, items)
            ChartControl.CrosshairEnabled = If(labelVisible, DefaultBoolean.False, CrosshairEnabled)
            ChartControl.ToolTipEnabled = If(labelVisible, DefaultBoolean.False, ToolTipEnabled)
        End Sub

        Private Sub UpdateControls()
            If OptionsLayoutControl Is Nothing Then Return
            OptionsLayoutControl.BeginUpdate()
            UpdateControlsCore()
            OptionsLayoutControl.EndUpdate()
        End Sub

        Private Sub TabbedView_PopupMenuShowing(ByVal sender As Object, ByVal e As XtraBars.Docking2010.Views.PopupMenuShowingEventArgs)
            e.Cancel = e.GetDocument() IsNot Nothing
        End Sub

        Protected Overridable Sub InitControls()
            If TabbedView IsNot Nothing Then AddHandler TabbedView.PopupMenuShowing, AddressOf TabbedView_PopupMenuShowing
            SetDataSources()
            If LabelAngleSpinEdit IsNot Nothing Then AddHandler LabelAngleSpinEdit.EditValueChanged, AddressOf LabelAngleSpinEditEditValueChanged
            If ValueAsPercentCheckEdit IsNot Nothing Then AddHandler ValueAsPercentCheckEdit.CheckedChanged, AddressOf ValueAsPercentCheckedEditCheckedChanged
        End Sub

        Protected Overridable Sub SetDataSources()
        End Sub

        Protected Overridable Sub UpdateControlsCore()
        End Sub

        Protected Overridable Sub OnDocumentActivatedCore(ByVal document As BaseDocument)
            UpdateControls()
            If AnimationEnabled Then activeChart.Animate()
        End Sub

        Protected Sub ApplySettingsToSeries(ByVal action As Action(Of SeriesBase))
            If String.IsNullOrEmpty(ChartControl.SeriesDataMember) AndAlso ChartControl.SeriesTemplate.DataAdapter Is Nothing Then
                For Each series As Series In ChartControl.Series
                    action(series)
                Next
            Else
                action(ChartControl.SeriesTemplate)
            End If
        End Sub

        Protected Sub ApplySettingsToSeriesView(ByVal viewAction As Action(Of SeriesViewBase))
            ApplySettingsToSeries(Sub(series) viewAction(series.View))
        End Sub

        Protected Sub SetEnabledLayoutControlItems(ByVal enabled As Boolean, ParamArray items As LayoutControlItem())
            For Each item As LayoutControlItem In items
                item.Enabled = enabled
            Next
        End Sub

        Protected Sub OnDocumentActivated(ByVal document As BaseDocument)
            If Not document.IsDockPanel Then Return
            activeChart = TryCast(CType(document.Control, DockPanel).ControlContainer.Controls(0), ChartControl)
            OnDocumentActivatedCore(document)
        End Sub

        Protected Sub SetLabelAngle(ByVal series As SeriesBase)
            Dim label As SeriesLabelBase = series.Label
            Dim labelAngle As Integer = CInt(LabelAngleSpinEdit.Value)
            Dim pointLabel As PointSeriesLabel = TryCast(label, PointSeriesLabel)
            If pointLabel IsNot Nothing Then
                pointLabel.Angle = labelAngle
                Return
            End If

            Dim rangeAreaLabel As RangeAreaSeriesLabel = TryCast(label, RangeAreaSeriesLabel)
            If rangeAreaLabel IsNot Nothing Then
                rangeAreaLabel.MinValueAngle = labelAngle
                rangeAreaLabel.MaxValueAngle = labelAngle
            End If
        End Sub

        Protected Sub SetLabelVisibility(ParamArray items As LayoutControlItem())
            OnLabelVisibilityChanged(items)
            Dim labelsVisibility As DefaultBoolean = CovertBoolToDefaultBoolean(LabelVisibleCheckEdit.Checked)
            ApplySettingsToSeries(Sub(series) series.LabelsVisibility = labelsVisibility)
        End Sub

        Protected Sub SetLabelVisibility(ByVal series As SeriesBase, ParamArray items As LayoutControlItem())
            OnLabelVisibilityChanged(items)
            series.LabelsVisibility = CovertBoolToDefaultBoolean(LabelVisibleCheckEdit.Checked)
        End Sub

        Protected Sub SetValueAsPercent(ByVal series As SeriesBase)
            Dim view As SeriesViewBase = series.View
            Dim valueAsPercent As Boolean = ValueAsPercentCheckEdit.Checked
            If TypeOf view Is FullStackedStepAreaSeriesView Then
                series.Label.TextPattern = If(valueAsPercent, PercentPattern, "{V:F0}")
            ElseIf IsPieView(view) Then
                series.Label.TextPattern = If(valueAsPercent, PiePercentPattern, "{A}: {V:F1}")
            ElseIf IsFunnelView(view) Then
                series.Label.TextPattern = If(valueAsPercent, FunnelPercentPattern, "{A}: {V:F0}")
            ElseIf IsSideBySideFullStackedBarView(view) Then
                series.Label.TextPattern = If(valueAsPercent, PercentPattern, "{V:0,,.0}M")
            ElseIf IsFullStackedBarView(view) Then
                series.Label.TextPattern = If(valueAsPercent, PercentPattern, "{V:F0}K")
            ElseIf IsFullStackedView(view) Then
                series.Label.TextPattern = If(valueAsPercent, PercentPattern, "${V}M")
            End If
        End Sub
    End Class
End Namespace
